### DISCUSSION TITLE
Unique Worker ID ?

### DISCUSSION THREAD

**Discussion by MuradAlvv** [Kludex/uvicorn#2459] [Ideas]
Hi,
Is it possible to get a unique integer ID for each worker when running Uvicorn in multi-process mode(--workers option) ? Instead of relying on PIDs, which can change, having an ID that ranges from 1 to the number of workers specified with --workers=WORKERS would be useful for instrumentation

**Comment by minmax** [comment]
~~`Worker.age` - unique auto-increment for each worker instance.~~

**Reply by MuradAlvv** [reply]
I didn't find anything about Worker.age in uvicorn documentation.

**Reply by minmax** [reply]
Sorry, I mixed up the topic with `UvicornWorker`

**Comment by whilenot-dev** [comment]
I would like to second the need for an unique identifier for each worker that is also consistent(!) across restarts.

I currently have a use case where I need to limit multiple resources for a synchronous REST API and that relates to the integrity of application state that is shared among all workers. This state is mutated by a background task that runs in each worker. I would need to provide an idempotent startup/shutdown routine to clean up any inconsistencies caused by a potential crash of any of my  workers (such a thing could happen with any unforeseen kill of the worker process). Without being able to differentiate between each worker in a way that is consistent across restarts I currently cannot make this routine idempotent.

As a workaround I implemented the background task to start with the acquisition of a lock that is shared among all worker - if the lock is aquired the background task starts, otherwise it's waiting (until the one worker with the aquired lock is killed/restarted). Unfortunately this limits my background tasks to a single instance for all workers instead of having one for each worker.

An implementation could as simple as passing the `_` (see [source](https://github.com/encode/uvicorn/blob/bfa754e21e2cc1d5b0d7cabf24933a6c3afc315e/uvicorn/supervisors/multiprocess.py#L123C13-L123C14)) as `process_num` to the worker process.

**Reply by whilenot-dev** [reply]
something like in [this draft](https://github.com/encode/uvicorn/compare/master...whilenot-dev:uvicorn:feat/unique-worker-id)?

**Reply by whilenot-dev** [reply]
moved draft to [a PR](https://github.com/encode/uvicorn/pull/2529)

**Comment by kamijin-fanta** [comment]
This feature is highly important as it is required for prometheus/client_python.
